TestRegistryEvents is flaky again

https://circleci.com/gh/weaveworks/scope/1115

```
--- FAIL: TestRegistryEvents-4 (0.11s)
	poll.go:25: /home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:209: 
		--- want
		+++ have
		@@ -1,5 +1,5 @@
		-([]docker.Container) (len=2 cap=2) {
		- (*docker_test.mockContainer)(0xc20803a350)({
		+([]docker.Container) (len=1 cap=1) {
		+ (*docker_test.mockContainer)(0xc2080e8020)({
		   c: (*docker.Container)(0x15ff2e0)({
		    ID: (string) (len=4) "ping",
		    Created: (time.Time) 0001-01-01 00:00:00 +0000 UTC,
		@@ -75,65 +75,6 @@
		    RestartCount: (int) 0,
		    AppArmorProfile: (string) ""
		   })
		- }),
		- (*docker_test.mockContainer)(0xc20803a358)({
		-  c: (*docker.Container)(0x15ff480)({
		-   ID: (string) (len=4) "wiff",
		-   Created: (time.Time) 0001-01-01 00:00:00 +0000 UTC,
		-   Path: (string) "",
		-   Args: ([]string) <nil>,
		-   Config: (*docker.Config)(0x1603280)({
		-    Hostname: (string) "",
		-    Domainname: (string) "",
		-    User: (string) "",
		-    Memory: (int64) 0,
		-    MemorySwap: (int64) 0,
		-    CPUShares: (int64) 0,
		-    CPUSet: (string) "",
		-    AttachStdin: (bool) false,
		-    AttachStdout: (bool) false,
		-    AttachStderr: (bool) false,
		-    PortSpecs: ([]string) <nil>,
		-    ExposedPorts: (map[docker.Port]struct {}) <nil>,
		-    Tty: (bool) false,
		-    OpenStdin: (bool) false,
		-    StdinOnce: (bool) false,
		-    Env: ([]string) <nil>,
		-    Cmd: ([]string) <nil>,
		-    DNS: ([]string) <nil>,
		-    Image: (string) "",
		-    Volumes: (map[string]struct {}) <nil>,
		-    VolumesFrom: (string) "",
		-    WorkingDir: (string) "",
		-    MacAddress: (string) "",
		-    Entrypoint: ([]string) <nil>,
		-    NetworkDisabled: (bool) false,
		-    SecurityOpts: ([]string) <nil>,
		-    OnBuild: ([]string) <nil>,
		-    Mounts: ([]docker.Mount) <nil>,
		-    Labels: (map[string]string) (len=2) {
		-     (string) (len=4) "foo1": (string) (len=4) "bar1",
		-     (string) (len=4) "foo2": (string) (len=4) "bar2"
		-    }
		-   }),
		-   State: (docker.State) Up 2562047h47m16.854775807s,
		-   Image: (string) (len=3) "baz",
		-   Node: (*docker.SwarmNode)(<nil>),
		-   NetworkSettings: (*docker.NetworkSettings)(<nil>),
		-   SysInitPath: (string) "",
		-   ResolvConfPath: (string) "",
		-   HostnamePath: (string) "",
		-   HostsPath: (string) "",
		-   LogPath: (string) "",
		-   Name: (string) (len=4) "waff",
		-   Driver: (string) "",
		-   Volumes: (map[string]string) <nil>,
		-   VolumesRW: (map[string]bool) <nil>,
		-   HostConfig: (*docker.HostConfig)(<nil>),
		-   ExecIDs: ([]string) <nil>,
		-   RestartCount: (int) 0,
		-   AppArmorProfile: (string) ""
		-  })
		  })
		 }
		 
FAIL
```

I'm also occasionally seeing this:

```
--- FAIL: TestRegistry-4 (0.02s)
	registry_test.go:193: 
		--- want
		+++ have
		@@ -1,16 +1,3 @@
		-([]*docker.APIImages) (len=1 cap=1) {
		- (*docker.APIImages)(0xcb22c0)({
		-  ID: (string) (len=3) "baz",
		-  RepoTags: ([]string) (len=2 cap=2) {
		-   (string) (len=4) "bang",
		-   (string) (len=10) "not-chosen"
		-  },
		-  Created: (int64) 0,
		-  Size: (int64) 0,
		-  VirtualSize: (int64) 0,
		-  ParentID: (string) "",
		-  RepoDigests: ([]string) <nil>,
		-  Labels: (map[string]string) <nil>
		- })
		+([]*docker.APIImages) {
		 }
		 
FAIL
```

--- a/probe/docker/registry_test.go
+++ b/probe/docker/registry_test.go
@@ -125,13 +125,17 @@ var (
 		},
 	}
 	apiContainer1 = client.APIContainers{ID: "ping"}
+	apiContainer2 = client.APIContainers{ID: "wiff"}
 	apiImage1     = client.APIImages{ID: "baz", RepoTags: []string{"bang", "not-chosen"}}
-	mockClient    = mockDockerClient{
+)
+
+func newMockClient() *mockDockerClient {
+	return &mockDockerClient{
 		apiContainers: []client.APIContainers{apiContainer1},
 		containers:    map[string]*client.Container{"ping": container1},
 		apiImages:     []client.APIImages{apiImage1},
 	}
-)
+}
 
 func setupStubs(mdc *mockDockerClient, f func()) {
 	oldDockerClient, oldNewContainer := docker.NewDockerClientStub, docker.NewContainerStub
@@ -172,8 +176,8 @@ func allImages(r docker.Registry) []*client.APIImages {
 }
 
 func TestRegistry(t *testing.T) {
-	mdc := mockClient // take a copy
-	setupStubs(&mdc, func() {
+	mdc := newMockClient()
+	setupStubs(mdc, func() {
 		registry, _ := docker.NewRegistry(10 * time.Second)
 		defer registry.Stop()
 		runtime.Gosched()
@@ -195,8 +199,8 @@ func TestRegistry(t *testing.T) {
 }
 
 func TestRegistryEvents(t *testing.T) {
-	mdc := mockClient // take a copy
-	setupStubs(&mdc, func() {
+	mdc := newMockClient()
+	setupStubs(mdc, func() {
 		registry, _ := docker.NewRegistry(10 * time.Second)
 		defer registry.Stop()
 		runtime.Gosched()
@@ -209,6 +213,7 @@ func TestRegistryEvents(t *testing.T) {
 
 		{
 			mdc.Lock()
+			mdc.apiContainers = []client.APIContainers{apiContainer1, apiContainer2}
 			mdc.containers["wiff"] = container2
 			mdc.Unlock()
 			mdc.send(&client.APIEvents{Status: docker.StartEvent, ID: "wiff"})
@@ -220,6 +225,7 @@ func TestRegistryEvents(t *testing.T) {
 
 		{
 			mdc.Lock()
+			mdc.apiContainers = []client.APIContainers{apiContainer1}
 			delete(mdc.containers, "wiff")
 			mdc.Unlock()
 			mdc.send(&client.APIEvents{Status: docker.DieEvent, ID: "wiff"})
@@ -231,6 +237,7 @@ func TestRegistryEvents(t *testing.T) {
 
 		{
 			mdc.Lock()
+			mdc.apiContainers = []client.APIContainers{}
 			delete(mdc.containers, "ping")
 			mdc.Unlock()
 			mdc.send(&client.APIEvents{Status: docker.DieEvent, ID: "ping"})

--- a/probe/docker/registry_test.go
+++ b/probe/docker/registry_test.go
@@ -1,7 +1,6 @@
 package docker_test
 
 import (
-	"reflect"
 	"runtime"
 	"sort"
 	"sync"
@@ -181,17 +180,16 @@ func TestRegistry(t *testing.T) {
 
 		{
 			want := []docker.Container{&mockContainer{container1}}
-			test.Poll(t, 10*time.Millisecond, want, func() interface{} {
+			test.Poll(t, 100*time.Millisecond, want, func() interface{} {
 				return allContainers(registry)
 			})
 		}
 
 		{
-			have := allImages(registry)
 			want := []*client.APIImages{&apiImage1}
-			if !reflect.DeepEqual(want, have) {
-				t.Errorf("%s", test.Diff(want, have))
-			}
+			test.Poll(t, 100*time.Millisecond, want, func() interface{} {
+				return allImages(registry)
+			})
 		}
 	})
 }

